bson/src/main/org/bson/RawBsonDocument.java

```diff
@@ -350,6 +350,7 @@ public BsonDocument clone() {
         return new RawBsonDocument(bytes.clone(), offset, length);
     }
 
+    @SuppressWarnings("unchecked")
     private BsonValue deserializeBsonValue(final BsonBinaryReader bsonReader) {
         return REGISTRY.get(getClassForBsonType(bsonReader.getCurrentBsonType())).decode(bsonReader, DecoderContext.builder().build());
```

Comment:
While this technique works, RawBsonDocumentCodec's reliance on BsonBinaryReader.pipe() means that the buffer has to be copied, which is not ideal.  Since RawBsonDocument is effectively read only we should be able to return a RawBsonDocument that shares the byte array of the parent RawBsonDocument.  So while not as clean it would be more efficient.  It would look something like:

    if (reader.currentType == DOCUMENT) {
        return new RawBsonDocument(bytes, 
                 /* get current position in reader */, 
                 /* get length from current position */);
    else {
       return REGISTRY....
    }

Reply:
I'll give it a go - but nested RawBsonDocuments in arrays will be BsonDocuments or contain copied bytes.

Reply:
True, but that can be fixed by adding a RawBsonArray class that complements RawBsonDocument.  That's worth doing though I'm fine deferring that to a subsequent review.